From c83ee6b933cf07d6c26a6462d86c0f6942bd33b9 Mon Sep 17 00:00:00 2001
From: Todd Poynor <toddpoynor@google.com>
Date: Mon, 27 Aug 2012 18:48:31 -0700
Subject: [PATCH 685/944] power: smb347: Allow disabling STAT interrupts

Allow STAT output to disable interrupt pulses and associated
configuration.  Can still use charging status output as an
interrupt if the irq_gpio is provided in platform data.

Based on code from hongmin.son <hongmin.son@samsung.com>

Change-Id: I036cd3c4b5b431f87d22c3a16a4ee0cecde432ce
Signed-off-by: hongmin.son <hongmin.son@samsung.com>
Signed-off-by: Todd Poynor <toddpoynor@google.com>
---
 drivers/power/smb347-charger.c       | 5 ++++-
 include/linux/power/smb347-charger.h | 1 +
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/power/smb347-charger.c b/drivers/power/smb347-charger.c
index 91bb6b0..188bc25 100644
--- a/drivers/power/smb347-charger.c
+++ b/drivers/power/smb347-charger.c
@@ -826,6 +826,8 @@ static int smb347_irq_init(struct smb347_charger *smb)
 		goto fail;
 
 	ret = request_threaded_irq(irq, NULL, smb347_interrupt,
+				   pdata->disable_stat_interrupts ?
+				   IRQF_TRIGGER_RISING | IRQF_ONESHOT :
 				   IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
 				   smb->client->name, smb);
 	if (ret < 0)
@@ -957,7 +959,8 @@ static int smb347_hw_init(struct smb347_charger *smb)
 
 	ret = smb347_update_online(smb);
 
-	if (smb->pdata->irq_gpio >= 0) {
+	if ((smb->pdata->irq_gpio >= 0) &&
+	    !smb->pdata->disable_stat_interrupts) {
 		/*
 		 * Configure the STAT output to be suitable for interrupts:
 		 * disable all other output (except interrupts) and make it
diff --git a/include/linux/power/smb347-charger.h b/include/linux/power/smb347-charger.h
index 32a155c..e9aab94 100644
--- a/include/linux/power/smb347-charger.h
+++ b/include/linux/power/smb347-charger.h
@@ -112,6 +112,7 @@ struct smb347_charger_platform_data {
 	bool		use_usb_otg;
 	bool		disable_automatic_recharge;
 	int		irq_gpio;
+	bool		disable_stat_interrupts;
 	enum smb347_chg_enable enable_control;
 	bool		usb_mode_pin_ctrl;
 	char		**supplied_to;
-- 
1.8.0

